### PR TITLE
Clarifications and "flexibility" fixes in Exemplar Specification

### LINKED ISSUE #2205: Incomplete example implementation of naive reservoir sampling algorithm
In [exemplar defaults](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/sdk.md#exemplar-defaults) the SimpleExemplarReservoir is defined as:

>SimpleExemplarReservoir This Exemplar reservoir MAY take a configuration parameter for the size of the reservoir pool. The reservoir will accept measurements using an equivalent of the [naive reservoir sampling algorithm](https://en.wikipedia.org/wiki/Reservoir_sampling)
>
>   ```
>  bucket = random_integer(0, num_measurements_seen)
>  if bucket < num_buckets then
>    reservoir[bucket] = measurement
>  end
>  ```

Reading the article, it seems like the implementation given here is incorrect or at least incomplete. For the first `n` elements where `n` is the size of the reservoir the exemplar should be offered to bucket `n-1`. Only after the reservoir is full should exemplars be sampled by generating the random number.

A more complete example might be:

```
bucket = num_measurements_seen if num_measurements_seen < reservoir.size else random_integer(0, num_measurements_seen)
if bucket < num_buckets then
	reservoir[bucket] = measurement
end
```

### LINKED ISSUE #2421: Exemplar clarifications
Doing a pass on the java exemplar implementation and found some places in the spec that need additional clarification IMO:

**More clearly define built-in exemplar filters**
 The spec says to "See [Defaults and Configuration](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/sdk-environment-variables.md#metrics-sdk-configuration) for built-in filters", which eventually leads to a list of `"none"`, `"all"`, and `"with_sampled_trace"`. These should be defined in the metric SDK document for improved clarify.

**Define how an [ExemplarFilter](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/sdk.md#exemplarfilter) gets configured in the SDK**
The Java SDK currently allows ExemplarFilter to be configured on a meter provider instance, which is presumably the right place, but that isn't explicitly stated anywhere. 

**Clarify whether views can configure exemplar reservoirs**
There's a [TODO](https://github.com/open-telemetry/opentelemetry-specification/blame/main/specification/metrics/sdk.md#L320-L322) that states "after we release the initial Stable version of Metrics SDK specification, we will explore how to allow configuring custom ExemplarReservoirs with the View API.", yet the view section [specifies](https://github.com/open-telemetry/opentelemetry-specification/blame/main/specification/metrics/sdk.md#L187-L189) that exemplar reservoirs can optionally be specified. 

Is the intention that views can be configured with variations of the [built-in](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/sdk.md#exemplar-defaults) reservoirs, but custom reservoirs are disallowed for the time being?

**Define the default size of `SimpleFixedSizeExemplarReservoir`**
If exemplars are enabled, `SimpleFixedSizeExemplarReservoir` is used by all aggregations except explicit bucket histogram, yet the default size of the reservoir is left unspecified.

### LINKED ISSUE #3669: Exemplar Reservoir Offer should be allowed to only accept dropped attributes
Currently the exemplar reservoir `Offer` method is defined as follows in [the specification](https://github.com/open-telemetry/opentelemetry-specification/blob/bc76c97525eef0f81550d939a3f87f44b182cef9/specification/metrics/sdk.md#exemplarreservoir):

> The "offer" method SHOULD accept measurements, including:
> 
> - [...]
> - The complete set of `Attributes` of the measurement.
> - [...]

However, the reservoirs are only expected to produce exemplars that hold the attributes that have been dropped from a measurement:

> `Exemplar`s MUST retain any attributes available in the measurement that
> are not preserved by aggregation or view configuration. Specifically, at a
> minimum, joining together attributes on an `Exemplar` with those available
> on its associated metric data point should result in the full set of attributes
> from the original sample measurement.

Requiring the `Offer` method receive all attributes when only the dropped will be preserved is inefficient if the implementation already knows the dropped attributes at the time `Offer` is called. It means the exemplar implementation will need to independently determine the dropped attributes.

### Background

The Go SDK filters measured attributes at the time they are recorded. This ensures that attribute filters can be utilized to restrict attributes for instrumentation with cardinality explosion and have it safely prevent run-away memory use.

In the process of filtering these measured attributes the calculation of the dropped attributes is trivial and also known. Based on the current specification, however, the Go implementation cannot simply provide these dropped attributes to the exemplar reservoir `Offer` method.

### Proposal

Update the specification to state that the attributes passed to `Offer` need to be either the dropped attributes or the complete set based on the implementation (but not both).

#### What about filtering with the kept attributes?

If the reservoir does not receive the complete set of attributes that are being recorded during a measurement it cannot filter based on these attributes. Therefore, this would reduce the functionality of the reservoir, right?

There are two points to be made here:

1. The exemplar filter would continue receive the complete set of attributes and could be used to filter on these attributes.
2. The reservoir could be made aware of the attributes that will be retained upon creation.

The second point is possible because all exemplar reservoir act as a reservoir for a distinct set of attributes. Given this one-to-one relationship when an exemplar reservoir is create it could be provided the set of attributes that will be retained and for each measurement it will be provided the set of attributes dropped. Therefore, it will have have access to the complete set of attributes.

If the functionality to filter on the complete set of attributes needs to be retained by the reservoir, an implementation can still provide this while still optimizing including the measurement process optimization of only receiving the dropped attributes.

### LINKED ISSUE #3674: Update default exemplar reservoir size for base-2 exponential histogram buckets
_Originally posted by @jsuereth in https://github.com/open-telemetry/opentelemetry-specification/issues/3670#issuecomment-1691744005_

I'd suggest limiting to a range of 5-20 exemplars (or some division of the exponential histogram's max size) purely from an overhead standpoint on tracking these points.  I don't have hard numbers to back this up, just some anecdotes.

Here's some VERY lazy math with naive assumptions:

### Explicit Bucket histograms
- Default 15 boundaries, 16 buckets
- 1:1 exemplar to bucket
- Likely storage per bucket count = 8 bytes
- Max, min, sum storage = 8*3 = 24 bytes
- Memory overhead without Exemplars: 16*8 + 8*3 = 152 bytes
- Likely storage per exemplar (best-case pointer to shared attributes, pointer to shared context + measurement + timestamp) = 32 bytes
- Memory head just for Exemplars = 512 bytes
- Total overhead / histogram = 664 bytes

### Exponential Bucket Histogram
- Default 320 buckets + overflow (Let's assume negative measurements don't exist)
- Likely storage per bucket count = ~2 bytes (assuming we're using "adapting integer" algorithm where we scale arrays as counts grow, and we're in a sweet spot of measurement-counts)
- max, min, sum storage = 8*3 bytes
- Memory overhead without exemplars - ~640 + 24 = 664 bytes

Given the key size differential, I don't think we should be spending quite as much memory overhead on exemplars.  Particularly because Exponential Buckets are intended to give us very high accuracy percentiles.  I think we're better of attempting to grab lower number but higher entropy exemplars.

### PR DESCRIPTION
**Pull request by jsuereth**
Fixes #2205
Fixes #3674 
Fixes #3669
Partially fixes #2421

## Changes

- Update exemplar algorithm to account for initial reservoir fill
- Update fixed-size defaults to account for memory contention / optimization in Java impl
- Set a default for exponential histogram aggregation
- Clarify that ExemplarFilter should be configured on MeterProvider
- Make it clear that ONE reservoir is create PER timeseries datapoint (not one reservoir per view or metric name).
- Allow flexibility in Reservoir `offer` definition based on feedback from Go impl.

* Related issues #3756

### CODE REVIEW

**Review by MrAlias — APPROVED** [review]

**Review by jack-berg — APPROVED** [review]

File: specification/metrics/sdk.md [hunk 1]
```diff
@@ -972,11 +973,21 @@ Using this ExemplarFilter is as good as disabling Exemplar feature.
 An ExemplarFilter which makes those measurements eligible for being an
 Exemplar, which are recorded in the context of a sampled parent span.
 
+#### Configuration
+
+The ExemplarFilter SHOULD be a configuration parameter of a `MeterProvider` for
+an SDK. The default value for the filter SHOULD follow the
```
Comment (cijothomas) [inline specification/metrics/sdk.md:979]: why not mention the default right here, instead of asking to follow env variable doc?

File: specification/metrics/sdk.md [hunk 2]
```diff
@@ -993,18 +1004,25 @@ span context and baggage can be inspected at this point.
 The "offer" method does not need to store all measurements it is given and
 MAY further sample beyond the `ExemplarFilter`.
 
+The "offer" method MAY accept a filtered subset of `Attributes` which diverge
```
Comment (cijothomas) [inline specification/metrics/sdk.md:1007]: "filtered subset of `Attributes`" -- what is this referring to?

For the below example:
view(mymetric, attributes=key1,key2)
histogram.Record(10, key1=a,key2=b,key3=c,key4=d)

The current wording says Reservoir's offer method should be given complete set of attributes, i.e key1,key2,key3,key4. The proposed wording in this PR is allowing implementations to instead provide the "filtered subset" only i.e key1,key2. OR is it key3,key4?

If it is latter (key3,key4), then what is the need for providing the Attributes associated with timeseries point? Reservoir already knows key3,key4, and the metricpoint would track key1,key2, so merging them produce the original, full set...

If it is former (key1,key2), how can it ever produce the key3,key4, as the actual timeseries will also track (key1,key2), and we lose key3,key4 forever....

**Review by dashpole — APPROVED** [review]

File: specification/metrics/sdk.md (hunk 2 above)
Comment (jsuereth) [reply specification/metrics/sdk.md:1007]: The reason this exists was a request from the Go SDK.

The TL;DR: is that while we use `SHOULD` in other places, this is allowing for an alternative option.  THat alternative is:

- You have the ExemplarReservoir *know* about all attributes when it's constructed.
- The `offer` method only takes the *difference* in attributes as an argument (i.e. the SDK will be doing diff of attributes for you).

I think this *first* bullet may be the piece you're missing, or is worded poorly here.

This is to allow:

- Sampling algorithms to use the full set of attributes
- SDK authors to alter how they optimise reservoirs.


In Java we do NOT want to diff ahead of time, because benchmarking shows that diffing the attributes only on collection reduces overhead of metric collection + exemplars.  I believe this is because it's not guaranteed that any given measurement that enters a reservoir is preserved and diffing only needs to occur during metric collection, which is, basically, once a minute or so.

File: specification/metrics/sdk.md (hunk 1 above)
Comment (jsuereth) [reply specification/metrics/sdk.md:979]: Since you have to look at the ENV variable spec, I'm not sure what the gain is, outside of potential mismatch between this file and the env variable file.

File: specification/metrics/sdk.md [hunk 3]
```diff
@@ -993,18 +1004,25 @@ span context and baggage can be inspected at this point.
...
 decide whether "collect" should also reset internal storage for delta temporal
 aggregation collection, or use a more optimal implementation.
 
 `Exemplar`s MUST retain any attributes available in the measurement that
-are not preserved by aggregation or view configuration. Specifically, at a
-minimum, joining together attributes on an `Exemplar` with those available
-on its associated metric data point should result in the full set of attributes
-from the original sample measurement.
+are not preserved by aggregation or view configuration for the associated
+timeseries. Specifically, at a minimum, joining together attributes on an
```
[the remark this answers is not in the document]
Comment (jsuereth) [reply specification/metrics/sdk.md:1023]: Yeah, this is too verbose, can remove the 'at a minimum' phrasing.

File: specification/metrics/sdk.md (hunk 1 above)
Comment (cijothomas) [reply specification/metrics/sdk.md:979]: Given ENV variables are optional in general, one does not *have to* look at them.

In other places in the spec, we list this directly instead of pointing to different doc, so I'd says its best to be consistent. eg: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/sdk.md#built-in-samplers

Also, env variables could move at different pace towards stability!

File: specification/metrics/sdk.md (hunk 2 above)
Comment (cijothomas) [reply specification/metrics/sdk.md:1007]: > In Java we do NOT want to diff ahead of time, because benchmarking shows that diffing the attributes only on collection reduces overhead of metric collection + exemplars

Agree. This is what .NET also does.

Comment (cijothomas) [reply specification/metrics/sdk.md:1007]: The wording is not clear anyway. @jsuereth Could you answer the question from the example I used above?

File: specification/metrics/sdk.md [hunk 4]
```diff
@@ -1038,8 +1066,9 @@ cycle. For the above example, that would mean that the `num_measurements_seen`
 count is reset every time the reservoir is collected.
 
 This Exemplar reservoir MAY take a configuration parameter for the size of the
-reservoir pool. If no size configuration is provided, the default size of `1`
-SHOULD be used.
+reservoir pool. If no size configuration is provided, the default size MAY be
+the number of possible concurrent threads (e.g. numer of CPUs) to help reduce
+contention. Otherwise, a default size of `1` SHOULD be used.
```
Comment (jmacd) [inline specification/metrics/sdk.md:1071]: Is this "pool" defined anywhere else? I would suggest removing the concept, as it appears to be not about the size of the sample but an operational detail?

File: specification/metrics/sdk.md [hunk 5]
```diff
@@ -1015,9 +1033,15 @@ The SDK SHOULD include two types of built-in exemplar reservoirs:
...
-`SimpleFixedSizeExemplarReservoir`.
+By default:
+
+- Explicit bucket histogram aggregation with more than 1 bucket will
+use `AlignedHistogramBucketExemplarReservoir`.
+- Base2 Exponential Histogram Aggregation SHOULD use a
+  `SimpleFixedSizeExemplarReservoir` with a reservoir equal to the
+  smaller of the maximum number of buckets configured on the aggregation or
+  twenty (e.g. `min(20, max_buckets)`).
+- All other aggregations will use `SimpleFixedSizeExemplarReservoir`.
```
Comment (jmacd) [inline specification/metrics/sdk.md:1045]: By the way, I began by reviewing the existing specification, then I began looking at how I'll update the [Lightstep metrics SDK](https://github.com/lightstep/otel-launcher-go/blob/main/lightstep/sdk/metric/README.md) to implement this specification.

One detail that I found that could potentially be improved in the spec, from this SDK's perspective, concerns the use of alternatives. 

The LS metrics SDK has a "two-stage" design in which aggregations are accumulated w/ delta temporality & full cardinality briefly before collection, at which point a second set of filtered aggregators is used. For such an SDK with this kind of internal pipeline, it's not possible to implement exactly the simple-fixed algorithm given in pseudocode below. In other words, there is not a "correct" merge function that can take two reservoirs calculated w/ simple-fixed logic and combine them such that the inputs receive probability proportional to size. However, with a weighted sample, I can overcome this problem.

In this setting, I would suggest the spec could be softened, maybe. Instead of requiring a simple fixed reservoir, we can require any uniform-weight approach that preserves uniform probability over the set of qualifying metric events. In a single-stage pipeline, the simple-fixed approach works. In a multi-stage pipeline, I will use https://github.com/lightstep/varopt (although there are several alternatives that will do this too).

File: specification/metrics/sdk.md (hunk 2 above)
Comment (jsuereth) [reply specification/metrics/sdk.md:1007]: > The current wording says Reservoir's offer method should be given complete set of attributes, i.e key1,key2,key3,key4. The proposed wording in this PR is allowing implementations to instead provide the "filtered subset" only i.e key1,key2. OR is it key3,key4?

This would be `key3,key4`.  

I'm not sure I understand your comment:

> If it is latter (key3,key4), then what is the need for providing the Attributes associated with timeseries point? Reservoir already knows key3,key4, and the metricpoint would track key1,key2, so merging them produce the original, full set...

Unless the Reservoir can see key3,key4 *and* key1,key2 in some fashion, there's no way to produce the original set.   This PR gives you two options:


1.  `offer` is given the full set `key1,key2,key3,key4`. The SDK (or the `collect` method) knows that `key1,key2` are on the timeseries point and does a diff *at collection time*.  This is how java works.
2. `offer` is given a diffed set (`key3`,`key4`), where the SDK is responsible for diffing the timeseries attributes vs. the full set on the measurement.  The construction/constructor of the ExemplarReservoir would need to receiver `key1,key2` *specifically to enable sampling* when `offer` is called.  


The tradeoff here is around ease of sampling and deferred "diff" of attributes, vs. ahead of time `diff` and ease of collection.   

I'll try to improve the wording a second time.

File: specification/metrics/sdk.md (hunk 5 above)
Comment (jsuereth) [reply specification/metrics/sdk.md:1045]: I'd be happy to move to var-opt over time.  Absolutely agree on limitations of current algorithm.

Not part of this PR, but in a discussion w/ @jack-berg we're looking at:

1. Exposing the ability to configure exemplar reservoir "factory" with aggregation.
2. This would allow exporters to define/refine default reservoir sampling for exemplars.

My expectation is that we expose the ability to play with different implementations.  When we find one that has similar performance to the naive sampling reservoir but with better traits (like higher variance of varopt) we can move to it.

If you have suggestions on how to be less proscriptive here, let me know.  Today all defaults are "SHOULD" to allow us to transition in the future.   Additionally, given how exemplars manifest, I don't think changing the default sampling algorithm (e.g. on ExponentionalHistograms) would break users in anyway.

File: specification/metrics/sdk.md (hunk 4 above)
Comment (jsuereth) [reply specification/metrics/sdk.md]: Removed

File: specification/metrics/sdk.md (hunk 2 above)
Comment (cijothomas) [reply specification/metrics/sdk.md:1007]: Understood! I think providing an example here would help especially since it is easy to confuse "filtered subset" for the ones remaining after filtering vs the ones that got removed during filtering.

Comment (jsuereth) [reply specification/metrics/sdk.md:1007]: Ok, updated, PTAL

**Review by jmacd — APPROVED** [review]

File: specification/metrics/sdk.md (hunk 5 above)
Comment (jmacd) [reply specification/metrics/sdk.md:1046]: Sounds good. The "SHOULD" is good enough -- any implementation that runs into this will be left with some choices, and we can work it out later. I'll work out a prototype anyway, thanks.

File: specification/metrics/sdk.md (hunk 1 above)
Comment (jsuereth) [reply specification/metrics/sdk.md]: Updated
